### PR TITLE
reconfigure dock clients on strut partial change

### PR DESCRIPTION
**Pull request by acrisci**
Strut partial tells i3 where to reserve space for i3bar. It is affected
by the `position` bar config directive.

### CODE REVIEW

**Comment by acrisci** [comment]
I'm trying to fix #1294

**Comment by acrisci** [comment]
This actually requires #1453 to work.

**Comment by acrisci** [comment]
Now the bar position will change when we `reload`.

**Comment by stapelberg** [comment]
I’ll review this once #1453 is in.

File: i3bar/src/xcb.c [hunk 1]
```diff
@@ -1462,6 +1462,51 @@ void realloc_sl_buffer(void) {
...
+    }
+    return xcb_change_property(xcb_connection,
+            XCB_PROP_MODE_REPLACE,
+            output->bar,
+            atoms[_NET_WM_STRUT_PARTIAL],
+            XCB_ATOM_CARDINAL,
+            32,
+            12,
+            &strut_partial);
+
```
Comment (stapelberg) [inline i3bar/src/xcb.c]: nit: stray newline

File: src/handlers.c [hunk 2]
```diff
@@ -1134,6 +1134,76 @@ static bool handle_class_change(void *data, xcb_connection_t *conn, uint8_t stat
     return true;
 }
 
+/*
+ * Handles the _NET_WM_STRUT_PARTIAL property for allocating space for dock clients.
+ *
+ */
+static bool handle_strut_partial_change(void *data, xcb_connection_t *conn, uint8_t state, xcb_window_t window,
```
Comment (stapelberg) [inline src/handlers.c]: Can you add a testcase for this please?

**Comment by acrisci** [comment]
I added a test, but it is failing. However, it seems to work fine under normal conditions with i3bar.

The issue is that the request for the strut partial property fails with error code 3 (which I think is BadWindow). However, the con is found in the tree and the window id seems to be valid.

Any ideas on what is going on with the test?

File: src/handlers.c [hunk 3]
```diff
@@ -1174,6 +1174,85 @@ static bool handle_class_change(void *data, xcb_connection_t *conn, uint8_t stat
...
+
+    window_update_strut_partial(con->window, prop);
+
+    /* we only handle this change for dock clients */
+    if (con->parent == NULL || con->parent->type != CT_DOCKAREA) {
+        return true;
+    }
+
+    Con *search_at = croot;
+    Output *output = get_output_containing(con->geometry.x, con->geometry.y);
```
Comment (stapelberg) [inline src/handlers.c]: Using `con_get_output()` seems more straight-forward than `get_output_containing()` here.

File: src/handlers.c [hunk 4]
```diff
@@ -1174,6 +1174,85 @@ static bool handle_class_change(void *data, xcb_connection_t *conn, uint8_t stat
...
+            con->window->dock = W_DOCK_TOP;
+        } else {
+            DLOG("geom->y = %d >= rect.height / 2 = %d, it is a bottom dock client\n",
+                    con->geometry.y, (search_at->rect.height / 2));
+            con->window->dock = W_DOCK_BOTTOM;
+        }
+    }
+
+    /* find the dockarea */
+    Con *dockarea = con_for_window(search_at, con->window, NULL);
```
Comment (stapelberg) [inline src/handlers.c]: `con_for_window` could return NULL. If you’re sure that it cannot under these circumstances, at least add an `assert`.

File: testcases/t/102-dock.t [hunk 5]
```diff
@@ -143,6 +143,21 @@ wait_for_map $window;
...
+    PROP_MODE_REPLACE,
+    $window->id,
+    $atomname->id,
+    $atomtype->id,
+    32,         # 32 bit integer
+    12,
+    pack('L12', 0, 0, 0, 16, 0, 0, 0, 0, 0, 0, 1280, 0)
+);
+
+@docked = get_dock_clients('bottom');
```
Comment (stapelberg) [inline testcases/t/102-dock.t]: You’re missing a `sync_with_i3;` before this line.

**Comment by stapelberg** [comment]
Even with the `sync_with_i3`, the testcase makes i3 crash sometimes. Valgrind reports invalid writes. I think your new code (or perhaps old code, but that’s rather unlikely) is corrupting memory somewhere.

**Comment by acrisci** [comment]
It works just fine outside of the test environment.

**Comment by stapelberg** [comment]
So? Memory corruption tends to have that feature (“works just fine on my machine”).

**Comment by acrisci** [comment]
Thanks for the hint about the sync in the testcases. That led me to the memory error and I fixed it.

I messed up a rebase. Do I need to close this and submit another pull request?

File: src/handlers.c (hunk 4 above)
Comment (acrisci) [inline src/handlers.c]: I think it should never be NULL because there is always a dock area below the output (or root) for the dock client.

**Comment by stapelberg** [comment]
Not sure what you mean with the messed up rebase? Can you not use interactive rebase and then just `git push -f`?

**Comment by acrisci** [comment]
It was showing a lot of recreated commits in the pull request for some reason. Maybe it was a github bug. I think that issue is resolved though and this should be ready.

**Pull request merged**